src/main/java/org/mariadb/jdbc/internal/failover/AbstractMastersListener.java

```diff
@@ -352,7 +416,28 @@ public FailLoop(Listener listener) {
             this.listener = listener;
         }
 
-        public void run() {
+        public void scheduleSelf(ScheduledExecutorService scheduler, boolean now) {
+            scheduledFuture = scheduler.scheduleWithFixedDelay(this, now ? 0 : 250, 250, 
+                                                               TimeUnit.MILLISECONDS);
+        }
+
+        // while scheduledFuture may be set from another thread
+        // it is expected only one thread will unschedule
+        @Override
+        public void unscheduleTask() {
+            while (true) {
+                if (scheduledFuture == null) {
+                    // not scheduled yet, should be a rare condition, just loop
+                    Thread.yield();
+                } else {
+                    scheduledFuture.cancel(false);
+                    scheduledFuture = null;
```

Comment:
I think I see the issue, though a return statement is also necessary, there seems to be an additional issue here.

The issue is that this was expected to only be invoked from stopFailover().  The reason this is necessary is because we assume at this point that a future _should_ exist, or will exist shortly.  stopFailover() is our thread safety to ensure that this assumption is accurate.

The issue is that this could also be invoked by "blockTillTerminated()".  A couple obvious solutions come to mind:
* we could make unscheduleTask defer to stopFailover(), and have stopFailover() invoke a new function with this same logic).
* We could move the AtomicReference into an AtomicBoolean inside this class, and have a final FailLoop instance that we defer to for scheduling.  This would require us to change our termination behavior to allow a task to be reset to a runnable state.

Let me know if you have a preference, the first option would have the less code change and would fit rather nicely.  The second option I think would be ultimately a bit cleaner, as it would reduce the dependence on the parent class.

For now I will probably just update the PR with the first option since it should be quick and easy to do, but let me know if you have a preference.